services/src/main/java/com/chatbot/services/PubSubPullSubscriber.java

```diff
@@ -0,0 +1,124 @@
+import com.google.api.gax.core.InstantiatingExecutorProvider;
+import com.google.cloud.pubsub.v1.AckReplyConsumer;
+import com.google.cloud.pubsub.v1.MessageReceiver;
+import com.google.cloud.pubsub.v1.Subscriber;
+import com.google.protobuf.Struct;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.PubsubMessage;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+
+// Annotating this class with @Component will launch the pull subscriber on application start up
+public class PubSubPullSubscriber {
+
+  @Autowired
+  private AsyncService asyncService;
+  private static String projectID;
+  private static String subscriptionID;
+  private static final String TRIGGER_EVENT_MESSAGE = "TriggerEvent";
+  private static final String SUGGEST_CATEGORY_CHANGE_EVENT = "SUGGEST_CATEGORY_CHANGE";
+  private static final String SUGGEST_IMAGE_UPLOAD_EVENT = "SUGGEST_IMAGE_UPLOAD";
+  private static Long maxOutstandingElements;
+  private static Long maxOutstandingBytes;
+
+  public PubSubPullSubscriber(
+      @Value("${pubsubConfig.maxOutstandingElements}") final String maxOutstandingElementsToSet,
+      @Value("${pubsubConfig.maxOutstandingBytes}") final String maxOutstandingBytesToSet)
+      throws InterruptedException {
+    projectID = System.getenv("projectID");
+    subscriptionID = System.getenv("subscriptionID");
+    maxOutstandingElements = Long.parseLong(maxOutstandingElementsToSet);
+    maxOutstandingBytes = Long.parseLong(maxOutstandingBytesToSet);
+    launchSubscriber();
+  }
+
+  private void launchSubscriber()
+      throws InterruptedException, IllegalArgumentException {
+    final ProjectSubscriptionName subscriptionName = ProjectSubscriptionName
+        .of(projectID, subscriptionID);
+    final MessageReceiver receiver =
+        (final PubsubMessage message, final AckReplyConsumer consumer) -> {
+      final String messageData = message.getData().toStringUtf8();
+      final Map<String, String> messageAttributesMap = message.getAttributesMap();
+      if(messageData.equals(TRIGGER_EVENT_MESSAGE)) {
+        final TriggerEventNotification triggerEventNotification = 
+            buildNotificationFromMessage(messageAttributesMap);
+        try {
+          asyncService.triggerEventHandler(triggerEventNotification);
+        } catch (final Exception e) {
+          e.printStackTrace();
+        }
+      } else {
+        throw new IllegalArgumentException("Unknown message received at subscriber");
+      }
+      consumer.ack();
+    };
+    final FlowControlSettings flowControlSettings = FlowControlSettings.newBuilder()
+        .setMaxOutstandingElementCount(maxOutstandingElements)
+        .setMaxOutstandingRequestBytes(maxOutstandingBytes)
+        .build();
+    final ExecutorProvider executorProvider = InstantiatingExecutorProvider.newBuilder()
+        .setExecutorThreadCount(4)
```

Comment:
pick this from application properties instead of assigning yourself.

Reply:
updated